File: src/common/web/data/entities/project/logbook/ProjectJobHistoryRecord.ts
import { type ConnectorInstanceID } from "../../connector/ConnectorInstance";
import { ProjectLogbookRecord, type RecordID } from "./ProjectLogbookRecord";

/**
 * Extended data of a project job history record.
 */
export type ProjectJobHistoryRecordExtData = Record<string, any>;

/**
 * Well-known IDs of extended data entries.
 */
export const enum ProjectJobHistoryRecordExtDataIDs {
    ExternalID = "external_id",
    DOI = "doi",
}

/**
 * A single record of a project's job history.
 *
 * @param connector_instance - The connector instance ID.
 * @param status - The success status (done or failed).
 * @param message - An optional message (usually in case of an error).
 */
export class ProjectJobHistoryRecord extends ProjectLogbookRecord {
    public readonly connector_instance: ConnectorInstanceID;

    public readonly success: boolean;
    public readonly message: string;

    public readonly ext_data: ProjectJobHistoryRecordExtData;

    public constructor(
        record: RecordID,
        timestamp: number,
        connectorInstance: ConnectorInstanceID,
        success: boolean,
        message: string,
        extData: ProjectJobHistoryRecordExtData = {},
    ) {
        super(record, timestamp);

        this.connector_instance = connectorInstance;

        this.success = success;
        this.message = message;

        this.ext_data = extData;
    }
}

File: src/common/py/data/entities/project/logbook/project_job_history_record.py
import enum
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from .project_logbook_record import ProjectLogbookRecord
from ...connector import ConnectorInstanceID

ProjectJobHistoryRecordExtData = typing.Dict[str, typing.Any]


class ProjectJobHistoryRecordExtDataIDs(enum.StrEnum):
    """
    Well-known IDs of extended data entries.
    """

    EXTERNAL_ID = "external_id"
    DOI = "doi"


@dataclass_json
@dataclass
class ProjectJobHistoryRecord(ProjectLogbookRecord):
    """
    A single record of a project's job history.

    Attributes:
        connector_instance: The connector instance ID.
        success: The success status (done or failed).
        message: An optional message (usually in case of an error).
        ext_data: Arbitrary extended data as a key-value dictionary.
    """

    connector_instance: ConnectorInstanceID = ""

    success: bool = True
    message: str = ""

    ext_data: ProjectJobHistoryRecordExtData = field(default_factory=dict)
